desktop-widgets/locationinformation.cpp

```diff
@@ -507,15 +506,14 @@ void DiveLocationLineEdit::focusOutEvent(QFocusEvent *ev)
 void DiveLocationLineEdit::itemActivated(const QModelIndex &index)
 {
 	QModelIndex idx = index;
-	if (index.column() == LocationInformationModel::UUID)
+	if (index.column() == LocationInformationModel::DIVESITE)
 		idx = index.model()->index(index.row(), LocationInformationModel::NAME);
 
-	QModelIndex uuidIndex = index.model()->index(index.row(), LocationInformationModel::UUID);
-	uint32_t uuid = uuidIndex.data().toInt();
-	currType = uuid == 1 ? NEW_DIVE_SITE : EXISTING_DIVE_SITE;
-	currUuid = uuid;
+	dive_site *ds = (dive_site *)index.model()->index(index.row(), LocationInformationModel::DIVESITE).data().value<void *>();
```

Comment:
Yuck. That is ugly as heck. And if I run into this line while chasing a bug, I'll claw my eyes out.
Can you do me a favor and add a comment what this actually DOES?

Reply:
Yes, I hate it too. We already have a few of those in the code base, notably when getting pointers to dives and trips out of the models. Therefore I thought let's continue using this idiom, as ugly as it is.

What happens:
When you get data out of a model, it is returned in a `QVariant`, which is kind of lame in that it doesn't support arbitrary types out-of-the-box. So what we did is cast to (void *) transport that in a `QVariant` and then cast back. Indeed ugly as hell.

This example here is even worse, because we access the model from the index and the make an index out of the index via the model and .... never mind.

Note that I didn't come up with that just added the void * casting - I'll try to come up with something more palatable.

====

That being said, the whole situation becomes even more crazy when QML is involved, because there, all QVariant<void *> seem to compare as equal - that revelation was quite a WTF-moment. What I therefore did was cast to uintptr_t(!)

====

One way out of this mess would perhaps be registering the types as meta-types with Qt's preprocessor directives. I believe that you can then put them in QVariants, but so far I haven't explored that option, as it also strikes me as not very elegant.

Does that clear it up somewhat?

Reply:
thanks for the explanations - every time I work with you on one of your PRs I learn something. Simply awesome.